src/main/java/com/minecolonies/colony/Colony.java

```diff
@@ -515,110 +452,169 @@ public void updateSubscribers()
+
             //  Send each type of update packet as appropriate:
             //      - To Subscribers if the data changes
             //      - To New Subscribers even if it hasn't changed
 
             //  ColonyView
-            if (isDirty || hasNewSubscribers)
+            sendColonyViewPackets(oldSubscribers, hasNewSubscribers);
+
+            // Permissions
+            sendPermissionsPackets(oldSubscribers, hasNewSubscribers);
+
+            //  Citizens
+            sendCitizenPackets(oldSubscribers, hasNewSubscribers);
+
+            //  Buildings
+            sendBuildingPackets(oldSubscribers, hasNewSubscribers);
+        }
+
+        isDirty = false;
+        isCitizensDirty = false;
+        isBuildingsDirty = false;
+        permissions.clearDirty();
+
+        buildings.values().forEach(AbstractBuilding::clearDirty);
+        citizens.values().forEach(CitizenData::clearDirty);
+    }
+
+    private void sendBuildingPackets(Set<EntityPlayerMP> oldSubscribers, boolean hasNewSubscribers)
+    {
+        if (isBuildingsDirty || hasNewSubscribers)
+        {
+            for (AbstractBuilding building : buildings.values())
             {
-                for (EntityPlayerMP player : subscribers)
+                if (building.isDirty() || hasNewSubscribers)
                 {
-                    boolean isNewSubscriber = !oldSubscribers.contains(player);
-                    if (isDirty || isNewSubscriber)
-                    {
-                        MineColonies.getNetwork().sendTo(new ColonyViewMessage(this, isNewSubscriber), player);
-                    }
+                    ColonyViewBuildingViewMessage msg = new ColonyViewBuildingViewMessage(building);
+
+                    subscribers.stream()
+                               .filter(player -> building.isDirty() || !oldSubscribers.contains(player))
+                               .forEach(player -> MineColonies.getNetwork().sendTo(msg, player));
                 }
             }
+        }
+    }
 
-            // Permissions
-            if(permissions.isDirty() || hasNewSubscribers)
+    private void sendCitizenPackets(Set<EntityPlayerMP> oldSubscribers, boolean hasNewSubscribers)
+    {
+        if (isCitizensDirty || hasNewSubscribers)
+        {
+            for (CitizenData citizen : citizens.values())
             {
-                subscribers
-                        .stream()
-                        .filter(player -> permissions.isDirty() || !oldSubscribers.contains(player)).forEach(player -> {
-                    Permissions.Rank rank = getPermissions().getRank(player);
-                    MineColonies.getNetwork().sendTo(new PermissionsMessage.View(this, rank), player);
-                });
+                if (citizen.isDirty() || hasNewSubscribers)
+                {
+                    ColonyViewCitizenViewMessage msg = new ColonyViewCitizenViewMessage(this, citizen);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'msg' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)